### ISSUE TITLE
Editing the table

### ISSUE THREAD

**Issue by deakjahn** [fiduswriter/simple-datatables#189]
I'm still trying but I thought I would ask in the meantime: is the [Editable plugin](https://github.com/Mobius1/Editable), prepared for the predecessor of this table library, still supported? Or is there any other plugin to make the table editable?

**Comment by deakjahn** [comment]
My actual problem is that I need editing, not just display. I don't even need the double click to edit mode functionality, even having INPUT and SELECT items all along would be fine. So I'm trying to accommodate it with your table because being simple and vanilla is just perfect for me.

**Comment by johanneswilm** [comment]
Ok, let me try to see if I can come up with a demo that shows how to do it.

**Comment by deakjahn** [comment]
In the meantime I started to populate the table with HTML like this:

        data: {
          headings: ['Index', 'Modifier', 'Type', 'Parameter', 'Default'],
          data: message.value.map(item => {
            return [item.index, select(item.modifier), input(item.type), input(item.parameter), input(item.value)];
          }),
        },

Those functions return the populated `<select>` and `<input>` tags. We'll see what this leads to. :-)

**Comment by deakjahn** [comment]
@johanneswilm I won't say I'm there yet. :-) `render` allows me to replace the simple display with editing controls but `export()` doesn't handle it, obviously, I need to write my own.

Besides, I think I found a bug: `render` is only called on the first setup, not when I add a new row with the rows API.

**Comment by deakjahn** [comment]
@johanneswilm So I turned the old `Editable` plugin into a class `DataTableEditable`. Its `constructor() and `init()``are ready to be called:

https://github.com/fiduswriter/Simple-DataTables/blob/ab183766ab934a36dcf34d0e397b5c1df85517b9/src/datatable.js#L106-L119

What's the secret sauce that makes it appear in `this[plugin]`?

**Comment by johanneswilm** [comment]
@deakjahn Where can I find your code? I can probably put it all together pretty quickly.

**Comment by deakjahn** [comment]
On my computer. :-) I'm trying to use it inside a Visual Studio Code extension plugin. That probably makes it a bit more complicated, I reckon.

**Comment by johanneswilm** [comment]
@deakjahn ok, if it's too complicated to share that code, I can try to do the conversion from scratch here. Sorry, I was trying to clean up some other things for version 5.0 first.

**Comment by deakjahn** [comment]
Before you do, could you simply confirm that you simply expect plugins to be ECMA modules, simply loaded as per normal, with the appropriate `ctor()` and `init()` to be called, and nothing else specific?

**Comment by deakjahn** [comment]
Because I could very well upload that file, I've just created a private repo here that I can add you to but I have no idea whether it runs at all. It's syntactically sane but I've yet to see a single line executed... :-)

**Comment by deakjahn** [comment]
Sent an invitation to it.

**Comment by deakjahn** [comment]
@johanneswilm Sent an invitation to it.

**Comment by johanneswilm** [comment]
Hey,
I'm 80% done getting it to work. The Editing plugin was written for the
never released Vanilla DataTables 2.0 so there are quite a lot of things to
change. Nevertheless a basic version should be ready soon.
Unfortunately, I cannot accept an invite to a closed repo, as that may
create licensing issues.

**Comment by deakjahn** [comment]
Just to work and discuss on it? When it's done, I'll send you the source code with whatever means you want it, MIT, of course, it's yours for inclusion.

**Comment by deakjahn** [comment]
@johanneswilm https://github.com/deakjahn/datatable-editable, I started an issue for discussion.

**Comment by johanneswilm** [comment]
This is what I have so far: https://github.com/fiduswriter/Simple-DataTables/commit/cf64b25b907efafd302c663a7905b7f432fe35b9

Editing cell content generally works, even though there probably is an issue with render. I have not looked at row editing yet.

**Comment by johanneswilm** [comment]
@deakjahn I would appreciate it if you could add any additions you want to make to the code in this repository as the old Vanilla DataTables and Simple DataTabkles differ considerably.

**Comment by deakjahn** [comment]
When I have finished it and if I finish it first, sure. But right now, it's not yet ready. I'll look into yours because mine happens to be the other way around, the row dialog appears but the in-place edit does not. :-)

**Comment by deakjahn** [comment]
@johanneswilm OK, that's a different approach, you just amalgamated the editing into the main code. I kept the plugin approach and made it work, mostly, not yet finished. :-)

Then I need to find out how I can test this in my own code now.

**Comment by johanneswilm** [comment]
@deakjahn I kept it in a separate function. You can see I made a new demo document (15). You can take a look at the source code of that to see how to use it.

**Comment by deakjahn** [comment]
Yep, I see all of that, the problem was more complicated than that: I needed to pack a UMD module from the code. That's what I have been using so far. While the scripting engine in VSCode is Node.js-based and fully compliant with modules, custom dialogs have to be implemented in a separate webview that isn't. And while the IDE has an integrated NPM, I didn't want to mess with it now, just to be safe, so I downloaded a separate portable NPM and will use that to pack your source. As soon as it's up and running, I'll report my fixes back.

**Comment by johanneswilm** [comment]
ok, super. I am done for now as basic row and cell editing is working, see https://fiduswriter.github.io/Simple-DataTables/15-editing/ .

There is still an issue with cells that use a render function of some kind. I think that issue may be affecting other parts as well. The way I plan on fixing it long-term is to change some of the internals, storing the table data in JavaScript as an array of arrays of content items apart from the dom of the table. I may even divide the dom up into a virtual and a real dom, using diffDOM [1] to apply changes to the virtual dom to the real one.  

If you find sensible ways to fix it for the editing part in the short term, then that is also very welcome.

[1] https://github.com/fiduswriter/diffDOM

**Comment by deakjahn** [comment]
@johanneswilm It pays to play safe. :-) The portable NPM didn't work, I ended up installing one but not on my primary dev machine and it sure would've installed into the same folder. OK, I'm in motion now again and I'm getting error messages like:

Uncaught TypeError: Cannot read properties of undefined (reading 'addEventListener')
    at g.bindEvents (datatable.js:1:19271)
    at g.render (datatable.js:1:15347)
    at g.init (datatable.js:1:12122)
    at new g (datatable.js:1:11882)
    at signature.js:51:14

Anyway, I'll go on now.

First question, is `enabled: true` really necessary in the editing options? After all, if you call `makeEditable()`, chances are you want to make it editable all right... I'll collect these, also implement my `excludeColumns` to disable editing without actually hiding columns and will upload those to a proper fork.

**Comment by johanneswilm** [comment]
> First question, is `enabled: true` really necessary in the editing options? 

I believe you are right. I only see it mentioned in the demo and removing it there doesn't seem to make any difference. So I've removed it now.

There are likely many more things that don't make sense. Feel free to change things as needed, even if it breaks backward compatibility. Noone has used the editing part in the past and after this is working there will need to be a version 5.0.x anyway, so this is the right moment to change things around.

**Comment by deakjahn** [comment]
@johanneswilm It started to work for me, too. A few small fixes were needed: https://github.com/deakjahn/Simple-DataTables/commit/b403152145a166224bb16159a4f65916f9eee609

Basically two, a `this.table` changed to `this.body`, the first didn't exist at all and gave me undefined. Second, `keyCode` is deprecated. I'll go on tomorrow with the testing (I don't know what time zone you're in but I'm in Europe, after midnight :-) ). It already edits and finally exports nicely.

Shall I PR these small changes or let's wait for more to come up? `editor.js` didn't change, I don't know why it picked up the parentheses for `rows()`, I reverted it.

**Comment by deakjahn** [comment]
Lest I forget:

- Repeated context menu Edit Cells leave the focus on the previous cell,
- After several such repeated menu Edit Cells (don't actually edit, just start editing in various cells), starting to click into cells will place the HTML `<input...>` into the editor field, not the actual content.

**Comment by johanneswilm** [comment]
@deakjahn OK, looks good so far and I merged it. I am in Sweden, so similar timezone. I can review again tomorrow or tonight.

> And a question before I do it: I'd suggest to change the `.dt-editor-` classes to `.dataTable-editor-` to keep it more in sync. What do you think?

I agree. Good idea to change it everywhere.

**Comment by deakjahn** [comment]
@johanneswilm Second round. Thoughts in the PR and in the commit descriptions.

**Comment by deakjahn** [comment]
Larger problems. :-) Adding a new row with `table.rows.add()` gets rid of all the previous edits.

**Comment by deakjahn** [comment]
@johanneswilm Only one major hurdle left for me now, the problem I mentioned above is part of a more general issue: rebuilding discards the edits. Consequently, operations that call `rebuild()` also do, like adding or removing rows or showing and hiding columns.

You must have a better overall view of how this could be fixed. :-)

**Comment by deakjahn** [comment]
This is my quick and dirty fix, changing the cell in two places in parallel: https://github.com/deakjahn/Simple-DataTables/commit/440d8aa2494875cdc7340d224ec3d72bfe965414

But this really needs your opinion before going any further with it. You know it inside out...

**Comment by johanneswilm** [comment]
Hmm, yeah let's do that for now. If I find time for it over the next few days, I think it will make sense to keep the entire structure as pure JS structure without dom elements in addition to everything and then use that as the source of truth for everything. Something like this:

```js
> console.log(dataTable.value)
{
  head: [
    'ID',
    'Name',
    'Age',
  ],
  body: [
    [1, 'Adrian', 23],
    [2, 'Pia', 24],
    [3, 'Hans', 75],
  ]
}
```

any change of value will then first change the contents inside of `dataTable.value` and then call the render function for that cell.

**Comment by deakjahn** [comment]
Most definitely. I would probably reformulate the `activeRows` as purely a view over the real data. For instance as an iterator backed by a generator that only yields visible cells.

But the rendering approach will also open up the possibility to use something other than a simple `<input>`, eg. a `<select>` for a dropdown.

**Comment by johanneswilm** [comment]
Yes, that may be a solution. Some tests with timing will have to be conducted. I assume that the reason for `dataTable.activeRows` being `dom` elements is that the previous maintainer saw some speed advantages to having those rather than rendering the creating new dom elements all the time. So it's possible that one needs to keep those around. In working with https://github.com/fiduswriter/diffDOM I discovered that a lot of time is lost if one frequently goes back and forth between dom and JS. So even doing searches on the "real data" may be faster than going through dom elements.

**Comment by deakjahn** [comment]
Considering that, up until now, it was only display, no change, requirements might have been different. Without being inside the package, just on superficial thought, my first approach certainly would have been not to hide the columns that way, actually removing and therefore needing to manipulate the active rows whenever there is a change, but simply rendering all and using CSS to hide what shouldn't be seen. But it might be possible that this is too naive an approach for real life cases, especially with many rows and many hidden columns, only the real experience woud tell.

**Comment by johanneswilm** [comment]
Yeah there is also a speed factor when the dom gets too large. So if you have 10000 or 100000 rows in a table inside of `document`, that will create an issue, even if most rows are hidden.

What I was surprised about with `diffDOM` is that I found out that there is even a speed issue with nodes that are held just in memory without being part of the `document`.

**Comment by johanneswilm** [comment]
Also, while we did not cover editing of cell contents, we did have issues with inserting rows that had cells that needed to be rendered. See for example https://github.com/fiduswriter/Simple-DataTables/issues/178

**Comment by deakjahn** [comment]
`render()` from the `columns` option is not called on subsequent operations, only during initial data setup. I also had this problem because my first column is a running index that needs to change to the ordinal index when rows are added or removed. I worked around it by simply correcting them myself after each modification but if rendering would be integral part of the display all the time, this would also become automatic: I can have a rende function that just returns `row.dataIndex + 1` and be done with it.

**Comment by deakjahn** [comment]
I did try to fix this, actually, by moving this:

https://github.com/fiduswriter/Simple-DataTables/blob/f91669a5732404c81a361e2d4c59a44d06c55f85/src/datatable.js#L663-L675

from its current place into the `rebuild()` function but it didn't work for some reason, so I gave up...

**Comment by deakjahn** [comment]
And now I know why it didn't work... The same reason as the one before: it only saves to `cell`, not both places. So I'll put this into a new PR temporarily, it works now. :-)

**Comment by markschmucker** [comment]
Hi guys, thanks for the work on this. I'd like to use the editing as demo'd at  https://fiduswriter.github.io/Simple-DataTables/15-editing/. A couple questions:

1. Adding or removing the class "dataTable-editor-editable" in Chrome dev tools seems to have no effect. Does it have to be specified when the table is created? If so, how can I specify it in the options when I create the datatable in javascript with "new simpleDatatables.DataTable(...)"?
2. Will the changes be pushed to jsDeliver or should I check out the latest from github?

**Comment by deakjahn** [comment]
1. As far as I can see, the CSS class doesn't appear used anywhere. I suspect Johannes added it in "just in case", if anybody ever wants to give it a different look.

**Comment by johanneswilm** [comment]
Hey @deakjahn and @markschmucker ,
I just refactored the entire library. I tried all the demos and it seems to all be working now. But please check and confirm.

**Comment by deakjahn** [comment]
@johanneswilm The trouble is that I used your module packed into a single JS and I already forgot how I made it. :-)

No, I recall it now. I had an npm on my laptop because I didn't want to mix it up with my dev machine. I webpacked it into an UMD module. Strangely enough, the same setup now gives errors, so I1ll have to look into it tomorrow. You don't want to include a pre-built `dist` into your project by any chance, do you? :-)

**Comment by johanneswilm** [comment]
@deakjahn There is a link to the upgrade guide in the README.md. Quite a lot has changed. Let me know if you have questions.

**Comment by johanneswilm** [comment]
@deakjahn There are prebuilt files that are used for the demos. Check the docs folder.

**Comment by deakjahn** [comment]
You're right. Then I actually built my own module back then because I needed to when I modified the source during testing, of course. I'll check it out tomorrow.

**Comment by deakjahn** [comment]
In the upgrade description, first sample, `rowIndex`, not `dataIndex`.

**Comment by johanneswilm** [comment]
@deakjahn You mean in this sentence:

> The dataIndex property on <tr>s is now called data-index and is always a string.

No, actually there used to be this custom variable called `dataIndex` on every `<tr>`. It could not be seen in the output HTML, so I didn't know about it until I read all the details of the documentation.

**Comment by johanneswilm** [comment]
@deakjahn Ah, just found it in the example. Thanks!

**Comment by deakjahn** [comment]
The table appears but I still have problems with editing. This situation might be specific but in VS Code where I need the table, using online styles is not allowed by the content security policy and it seems that you now use those while editing. More precisely, I can disable safety as per https://content-security-policy.com/examples/allow-inline-style/ but that's not nice. We would need a way to specify a nonce, unless you can simply move everything to classes and just toggle the classes. I didn't yet check the source to see what causes this.

**Comment by johanneswilm** [comment]
Where are we using inline styles? If there are just a few cases, we may be able to eliminate those.

**Comment by deakjahn** [comment]
Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self' 'self' https://*.vscode-cdn.net". Either the 'unsafe-inline' keyword, a hash ('sha256-jM21DycoYaDyo/303bHCFM4rD1vTQ6yCYKhewBxflN4='), or a nonce ('nonce-...') is required to enable inline execution. Note that hashes do not apply to event handlers, style attributes and javascript: navigations unless the 'unsafe-hashes' keyword is present.

And it says umd.js line 3, which doesn't really help us. :-)

**Comment by johanneswilm** [comment]
Just quickly browsing through the sources... I don't see any added instances of that. I do see a few places where we are dynamically creating CSS for distances like in the example above. I could spot 1 -3 cases where we can use a class instead.

**Comment by deakjahn** [comment]
Basically, the table appears all right for me but nothing happens when I click on it, and I get a few error messages like this. I don't yet know it this is all that makes it stop or if I still have other issues with the upgrade process.

**Comment by johanneswilm** [comment]
> Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self' 'self' https://*.vscode-cdn.net". Either the 'unsafe-inline' keyword, a hash ('sha256-jM21DycoYaDyo/303bHCFM4rD1vTQ6yCYKhewBxflN4='), or a nonce ('nonce-...') is required to enable inline execution. Note that hashes do not apply to event handlers, style attributes and javascript: navigations unless the 'unsafe-hashes' keyword is present.


I am getting old. I don't know what any of that means. :(

**Comment by johanneswilm** [comment]
@deakjahn try maybe without the editing plugin and just write your own code where you listen to a click event on `dataTable.dom` filtering with `event.target` and then show the value of something like `dataTable.data.data[0]` in an alert if the user clicked the right place. If that shows you the first row, then it should be possible... although maybe not with that plugin.

**Comment by deakjahn** [comment]
Basically, I have some settings in the head of my HTML that specifies what I want to allow and where from: images, scripts, styles and fonts are all under strict surveillance and the system only allows them to be loaded from predetermined locations. Either I use hashes when the system makes sure that I use the same value in both places, both in the HTML head and when I actually refer to an inline style, to make sure it's me who asks it, or I relax the surveillance by allowing inline styles to come from anywhere. This certainly doesn't normally happen on a regular website (or, who knows), but systems where the app actually runs in an embedded browser (VS Code itself is an application written in Node.js and Typescript and whatnot, pure black magic), well, it probably is quite common these days.

**Comment by johanneswilm** [comment]
ok, understood. The main places where we specify style properties from JavaScript is when we measure the width/size of some X and then apply a margin/width/etc. of size X to some other item so that they match up in the output. So we are not doing more than what we were doing in version 5 as well. Also, all the class names are now configurable.

**Comment by deakjahn** [comment]
Anyway, let's pretend it's my specific problem if it isn't easy to spot, now I added the unsafe specification, it allows them now, we'll see.

What about the rest? Am I allowed to handle `datatable.selectrow` as before? To change rows by simply swapping `grid.data[row1]` and `grid.data[row2]`?

**Comment by johanneswilm** [comment]
dataTable.rows.selectRow(dataIndex: int) <- note that it is an integer now.

**Comment by deakjahn** [comment]
No, not to select, to be informed when it's clicked on. The event.

**Comment by johanneswilm** [comment]
oh, yeah, the events should be the same - in a few cases with a little extra information.

**Comment by deakjahn** [comment]
Two swap rows, I probably need to extract the data of both, and call UpdateRow on the with the sapped data, is that right? Although I can't yet see what the `select` argument expects...

**Comment by johanneswilm** [comment]
All the data is in it's basically original format stores inside of `dataTable.data`. You can make any change in there you like. When you are done and you want the dom to update, you need to somehow call the `dataTable.renderTable()` method. The way to do that seems to be:

* If the widths of the columns haven't changed, run `dataTable.update()`

* If the widths could have changed, call something like `dataTable.update(false) && dataTable.fixColumns()`.

**Comment by johanneswilm** [comment]
I have only optimized the rendering of the main table. The pager and other things around it still work in the same way it used to do.

**Comment by deakjahn** [comment]
I don't have pages. :-)

This seems to work,with the usual problem that the changes are not concurrent, so it only works then modifying downwards, otherwise I have to swap the two updateRow() lines:

    const data1 = grid.data.data[row1];
    const data2 = grid.data.data[row2];
    grid.rows.updateRow([row1], data2);
    grid.rows.updateRow([row2], data1);
    grid.rows.setCursor(row2);
    grid.update();

**Comment by johanneswilm** [comment]
That should work, yes. But it's probably a bit slower as it's checking the data as if it was data that came from the outside. 

I would just treat `grid.data.data` like an array of all your rows and each row is an array of cells. put them together however you want and then render the table when you are done. `dataTable.fixColumns()` renders the table twice - once to measure the width of content and a second time to actually do the real rendering. So if that is not something you need, I would avoid calling it.

**Comment by deakjahn** [comment]
And I can't immediately see how that works, that's not my code. There's no reaction, just nothing happens, no error message. Double and single clicks are still supported, aren't they?

Oh, no problems, I have just a few lines of table at most, five columns, all local.

**Comment by johanneswilm** [comment]
Yeah, the only thing to keep in mind is that if you place event listeners on anything below `dataTable.dom`, those could randomly be removed by the renderer by accident. So I would not do that.

**Comment by deakjahn** [comment]
Yes, indeed, swapping directly works just as well, as earlier.

**Comment by deakjahn** [comment]
So. If I *don't* have `clickEvent: 'click`, the dialog pops up, no inline editing. If I do have it, it does nothing, neither inline nor dialog.

**Comment by deakjahn** [comment]
Sorry to say, demo 15 doesn't work for me even simply in a broswer, either. :-)

**Comment by johanneswilm** [comment]
@deakjahn That's odd. It's working for me in both Brave and Firefox on Linux. Which browser are you using?

**Comment by johanneswilm** [comment]
So this page: https://fiduswriter.github.io/simple-datatables/15-editing/ ? If you double click on a cell or right click on a row, what happens?

**Comment by deakjahn** [comment]
It works here, but only to bring up the dialog, not inline editing.

Vivaldi, Chromium engine.

**Comment by johanneswilm** [comment]
Ah no, there is no inline editing there. Mainly that is because you are the only user of that feature and it wouldn't have worked with the new version as it's now employing dom diffing through diffDOM. But I can think of a few ways we could get it back:

1. Create a modal that exactly covers the cell you are editing so that it appears as if it's inline editing. Difficulty would be how to get the positioning right. 

2. Copy the virtual dom, replace the contents of the cell with an input field in the copy, render the table with it. Make sure to shut off other event listeners that otherwise could render the table again while one is editing it. Difficulty would be that this is quite involved. But still probably just 2-3 hours altogether.

3. Temporarily turn off all the event listeners that may cause, then just manipulate the dom manually. When one is done editing, put everything back in place as it was (dom and event listeners), then rerender the table with the new data. The problem is it is a bit hacky and it will blow up if things are put back just slightly differently.

**Comment by deakjahn** [comment]
OK, I'll try. How come not everybody wants it? Most tables work that way today. :-)

**Comment by jawobar** [comment]
I want it! :) I actually was using a working example with inline editing working and found it is not working anymore after the refactoring. So there are at least two users of that feature now ;)

**Comment by johanneswilm** [comment]
@jawobar @deakjahn Ok, then I think you should probably be able to get it to work again between the two of you. I did most of the boring part of the conversion job of the editing plugin already. The only thing I left out was the inline editing. I'm here if you have any question. I'm looking forward to a PR!

**Comment by johanneswilm** [comment]
@jawobar @deakjahn Version 7.0.0-rc.2 allows inline editing again - also of rows.

**Comment by jawobar** [comment]
Awesome! Demo looks really good. Exactly what I need. Will make some good use of it. Thanks @johanneswilm !
